File: operonx/core/tracing/pipeline.py
"""TracePipeline — owns the event buffer, runs processors, dispatches to exporters.

Bound per ``engine.start()`` call. The execution model
(``docs/TRACING_REDESIGN_PLAN.md`` §3.7) constrains what runs where:

- ``_push(event)`` is sync, O(1) — called from ``EventEmitter.emit``
- ``flush_strategy.should_flush(event, buffered)`` is checked per-push;
  on True, ``_schedule_flush()`` fires off ``asyncio.create_task(self._run_flush())``
  so the current op continues without waiting
- ``_run_flush`` runs the processor chain on the event loop (pure-Python work,
  GIL-bound — a thread pool gives no benefit)
- Each exporter's ``export()`` runs via ``loop.run_in_executor(...)`` so
  blocking I/O (HTTP POST, file write) doesn't stall the main loop

Phase T1 ships the skeleton: buffer + flush trigger plumbing only. Processor
chain and exporter dispatch are stubbed; T2 fills them in alongside
re-implementations of the existing tracers.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Callable, Dict, Iterable, List, Optional, Protocol

from operonx.core.tracing.emitter import EventEmitter
from operonx.core.tracing.events import EventKind, TraceEvent

LOGGER = logging.getLogger("operonx.tracing")


Processor = Callable[[Iterable[TraceEvent]], Iterable[TraceEvent]]
"""A pure stream transform. See ``docs/TRACING_REDESIGN_PLAN.md`` §3.3."""


class Exporter(Protocol):
    """Renders + sends a processed event list to a backend.
    See ``docs/TRACING_REDESIGN_PLAN.md`` §3.4."""

    def export(
        self,
        events: List[TraceEvent],
        request_id: str,
        metadata: Dict[str, Any],
    ) -> None: ...


class FlushStrategy(Protocol):
    """Decides whether to flush after each emitted event.
    See ``docs/TRACING_REDESIGN_PLAN.md`` §7."""

    def should_flush(
        self,
        event: TraceEvent,
        buffered: int,
    ) -> bool: ...


class AtScheduledExit:
    """Default — flush only when the scheduler task ends (current behavior)."""

    def should_flush(self, event: TraceEvent, buffered: int) -> bool:
        return False


class FlushOnSize:
    """Flush whenever the buffer hits ``max_events``. Bounded memory."""

    def __init__(self, max_events: int) -> None:
        self.max_events = max_events

    def should_flush(self, event: TraceEvent, buffered: int) -> bool:
        return buffered >= self.max_events


class FlushOnGroupEnd:
    """Flush at every ``GROUP_END`` whose name matches ``pattern``.

    Pattern is a glob — ``"turn-*"`` matches ``"turn-0"``, ``"turn-1"``, etc.
    Used by callbot for per-turn live export to Langfuse.
    """

    def __init__(self, group: str) -> None:
        self.pattern = group

    def should_flush(self, event: TraceEvent, buffered: int) -> bool:
        if event.kind is not EventKind.GROUP_END:
            return False
        from fnmatch import fnmatchcase
        name = event.payload.get("name") or ""
        return fnmatchcase(name, self.pattern)


class TracePipeline:
    """Owns the per-call event buffer + processor chain + exporter dispatch.

    Construction is cheap; the heavy work happens in ``flush()``. One
    pipeline per ``engine.start()`` call — concurrent calls have isolated
    buffers and emitters via the ``_current_emitter_var`` ContextVar.

    Phase T1: ``flush()`` is a stub (no processors run yet). Wire-up to
    base.py / task_scheduler.py emit sites lands in the same phase. The
    actual processor + exporter machinery is built in T2.
    """

    def __init__(
        self,
        processors: Optional[List[Processor]] = None,
        exporters: Optional[List[Exporter]] = None,
        flush_strategy: Optional[FlushStrategy] = None,
        max_buffered_events: int = 100_000,
    ) -> None:
        self.processors: List[Processor] = list(processors or ())
        self.exporters: List[Exporter] = list(exporters or ())
        self.flush_strategy: FlushStrategy = flush_strategy or AtScheduledExit()
        self.max_buffered_events = max_buffered_events

        self._buffer: List[TraceEvent] = []
        self._overflow_warned = False
        self._pending_flush: Optional[asyncio.Task] = None
        # Per-call metadata captured at emitter() time. Forwarded to each
        # exporter as the ``metadata`` arg of ``export(...)``. Keys mirror
        # the legacy LangfuseTracer.flush() shape: workflow_name, user_id,
        # session_id, tags. Exporters pick what they need.
        self._metadata: Dict[str, Any] = {}

    # ------------------------------------------------------------------
    # Per-call binding
    # ------------------------------------------------------------------

    def emitter(
        self,
        request_id: str,
        *,
        workflow_name: Optional[str] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ) -> EventEmitter:
        """Construct a fresh emitter bound to this pipeline + request_id.

        Caller (engine.start) is responsible for setting ``_current_emitter_var``
        so ops + scheduler can find it.

        The keyword args are stashed on the pipeline and forwarded to each
        exporter's ``export(events, request_id, metadata)`` call. They mirror
        the legacy ``LangfuseTracer.flush(trace_data)`` shape.
        """
        self._metadata = {
            "workflow_name": workflow_name,
            "user_id": user_id,
            "session_id": session_id,
            "tags": tags,
        }
        return EventEmitter(pipeline=self, request_id=request_id)

    # ------------------------------------------------------------------
    # Internal — emit path
    # ------------------------------------------------------------------

    def _push(self, event: TraceEvent) -> None:
        """Append + check flush trigger. Called by ``EventEmitter.emit``.

        Hot path: must stay O(1). The flush trigger check is a bool
        predicate. On True, fires an ``asyncio.create_task`` so the current
        op continues without waiting.
        """
        self._buffer.append(event)

        # Bounded buffer — overflow → drop oldest + WARN once.
        if len(self._buffer) > self.max_buffered_events:
            drop_count = len(self._buffer) - self.max_buffered_events
            del self._buffer[:drop_count]
            if not self._overflow_warned:
                LOGGER.warning(
                    "trace buffer overflow at request_id=%s — dropped %d oldest "
                    "events (max_buffered_events=%d). Consider FlushOnGroupEnd "
                    "or FlushOnSize for long-running calls.",
                    event.request_id, drop_count, self.max_buffered_events,
                )
                self._overflow_warned = True

        # Flush trigger — non-blocking schedule.
        if self.flush_strategy.should_flush(event, len(self._buffer)):
            self._schedule_flush(partial=True)

    def _schedule_flush(self, partial: bool) -> None:
        """Fire-and-forget flush task. Returns immediately.

        If a previous flush is still pending, skip — the in-flight task
        will pick up everything currently buffered when it gets to drain.
        """
        if self._pending_flush is not None and not self._pending_flush.done():
            return
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            # No loop running — likely a sync test or shutdown path.
            # Fall back to synchronous flush so events aren't lost.
            self._flush_sync(partial=partial)
            return
        self._pending_flush = loop.create_task(self._run_flush(partial=partial))

    # ------------------------------------------------------------------
    # Flush — runs processors + exporter dispatch
    # ------------------------------------------------------------------

    async def flush(self, partial: bool = False) -> None:
        """Drain the buffer through processors + exporters.

        Called explicitly by ``engine.start``'s finally (final flush) or
        indirectly via ``_schedule_flush`` (partial flushes triggered by
        flush_strategy).
        """
        await self._run_flush(partial=partial)

    async def _run_flush(self, partial: bool) -> None:
        """Run processors on the asyncio loop, exporters in a thread pool."""
        if not self._buffer:
            return

        events: List[TraceEvent] = self._buffer
        self._buffer = []

        # Processors run on the event loop — pure-Python CPU work, GIL-bound.
        try:
            stream: Iterable[TraceEvent] = events
            for proc in self.processors:
                stream = proc(stream)
            processed = list(stream)
        except Exception:
            LOGGER.exception("processor chain failed; dropping batch")
            return

        if not processed:
            return

        # Each exporter runs in the thread pool — HTTP / file I/O.
        loop = asyncio.get_running_loop()
        request_id = processed[0].request_id
        metadata: Dict[str, Any] = {**self._metadata, "partial": partial}

        for exporter in self.exporters:
            try:
                await loop.run_in_executor(
                    None, exporter.export, processed, request_id, metadata,
                )
            except Exception:
                LOGGER.exception(
                    "exporter %s failed for request_id=%s",
                    type(exporter).__name__, request_id,
                )

    def _flush_sync(self, partial: bool) -> None:
        """Synchronous fallback when no event loop is running.

        Used at shutdown or in non-async tests. Processors run inline;
        exporters run inline (no executor). Failures logged, not raised.
        """
        if not self._buffer:
            return
        events = self._buffer
        self._buffer = []
        try:
            stream: Iterable[TraceEvent] = events
            for proc in self.processors:
                stream = proc(stream)
            processed = list(stream)
        except Exception:
            LOGGER.exception("processor chain failed (sync); dropping batch")
            return
        if not processed:
            return
        request_id = processed[0].request_id
        metadata = {**self._metadata, "partial": partial}
        for exporter in self.exporters:
            try:
                exporter.export(processed, request_id, metadata)
            except Exception:
                LOGGER.exception(
                    "exporter %s failed (sync) for request_id=%s",
                    type(exporter).__name__, request_id,
                )

File: rust/operonx/src/core/tracing/pipeline.rs
//! `TracePipeline` — owns the event buffer, runs processors, dispatches to exporters.
//!
//! Mirrors Python [`operonx/core/tracing/pipeline.py`](../../../../../operonx/core/tracing/pipeline.py).
//!
//! Per the execution-model rules (§3.7), the `push` path is sync + O(1)
//! — events land on a bounded buffer and `_run_flush` runs the processor
//! chain + exporter dispatch off the hot path. Exporters perform blocking
//! I/O so the flush body uses `tokio::task::spawn_blocking` to keep the
//! main runtime responsive.

use std::sync::Arc;

use async_trait::async_trait;
use parking_lot::Mutex;
use serde_json::Value;
use tracing::warn;

use super::events::TraceEvent;

/// A pure stream transform applied during flush.
pub trait Processor: Send + Sync {
    fn name(&self) -> &'static str {
        "unnamed"
    }
    fn process(&self, events: Vec<TraceEvent>) -> Vec<TraceEvent>;
}

/// Renders + sends a processed event list to a backend.
#[async_trait]
pub trait Exporter: Send + Sync {
    fn name(&self) -> &'static str {
        "unnamed"
    }
    async fn export(
        &self,
        events: Vec<TraceEvent>,
        request_id: String,
        metadata: ExportMetadata,
    );
}

/// Decides whether to flush after each emitted event.
pub trait FlushStrategy: Send + Sync {
    fn should_flush(&self, event: &TraceEvent, buffered: usize) -> bool;
}

/// Default — flush only when the engine explicitly drains.
pub struct AtScheduledExit;

impl FlushStrategy for AtScheduledExit {
    fn should_flush(&self, _event: &TraceEvent, _buffered: usize) -> bool {
        false
    }
}

/// Flush whenever the buffer hits `max_events`.
pub struct FlushOnSize {
    pub max_events: usize,
}

impl FlushStrategy for FlushOnSize {
    fn should_flush(&self, _event: &TraceEvent, buffered: usize) -> bool {
        buffered >= self.max_events
    }
}

/// Per-call metadata forwarded to each exporter. Mirrors Python's pipeline
/// `_metadata` dict (`workflow_name`, `user_id`, `session_id`, `tags`,
/// `partial`).
#[derive(Debug, Clone, Default)]
pub struct ExportMetadata {
    pub workflow_name: Option<String>,
    pub user_id: Option<String>,
    pub session_id: Option<String>,
    pub tags: Vec<String>,
    pub partial: bool,
}

/// Owns the event buffer + processor chain + exporter list. One per
/// `engine.start()` call; the emitter wraps it in an `Arc` and pushes
/// events onto it from anywhere.
pub struct TracePipeline {
    processors: Vec<Arc<dyn Processor>>,
    exporters: Vec<Arc<dyn Exporter>>,
    flush_strategy: Arc<dyn FlushStrategy>,
    max_buffered_events: usize,
    buffer: Mutex<Vec<TraceEvent>>,
    overflow_warned: Mutex<bool>,
    metadata: Mutex<ExportMetadata>,
}

impl TracePipeline {
    pub fn new() -> Self {
        Self {
            processors: Vec::new(),
            exporters: Vec::new(),
            flush_strategy: Arc::new(AtScheduledExit),
            max_buffered_events: 100_000,
            buffer: Mutex::new(Vec::new()),
            overflow_warned: Mutex::new(false),
            metadata: Mutex::new(ExportMetadata::default()),
        }
    }

    pub fn builder() -> TracePipelineBuilder {
        TracePipelineBuilder::default()
    }

    /// Set the per-call metadata. Called by the engine before the scheduler
    /// runs so exporters see the right `workflow_name` / `user_id` / etc.
    pub fn set_metadata(&self, meta: ExportMetadata) {
        *self.metadata.lock() = meta;
    }

    /// Append an event. Matches Python's `_push` — hot path, O(1) + bounded
    /// buffer overflow handling.
    pub fn push(&self, event: TraceEvent) {
        let mut buf = self.buffer.lock();
        buf.push(event.clone());
        if buf.len() > self.max_buffered_events {
            let drop_count = buf.len() - self.max_buffered_events;
            buf.drain(..drop_count);
            let mut warned = self.overflow_warned.lock();
            if !*warned {
                warn!(
                    "trace buffer overflow at request_id={} — dropped {} oldest events \
                     (max_buffered_events={}). Consider FlushOnSize for long-running calls.",
                    event.request_id, drop_count, self.max_buffered_events
                );
                *warned = true;
            }
        }
        let buffered_len = buf.len();
        let should_flush = self.flush_strategy.should_flush(&event, buffered_len);
        drop(buf);
        if should_flush {
            // Best-effort — spawn into the current runtime if one is
            // available; otherwise the caller is responsible for invoking
            // `flush().await` at the boundary.
            if let Ok(handle) = tokio::runtime::Handle::try_current() {
                let pipeline = self as *const Self;
                // SAFETY: TracePipeline is owned by an Arc on the call
                // side; the engine keeps it alive for the full request.
                // The spawned task gets a Send pointer via the Arc shape
                // we use below — implemented as `flush_in_background`.
                let _ = pipeline;
                let _ = handle;
                // No-op here: actual scheduling requires the Arc, see
                // `flush_in_background()` companion below.
            }
        }
    }

    /// Drain the buffer (no processing). Test-only helper — production
    /// callers use `flush()` which applies processors + exporters.
    #[doc(hidden)]
    pub fn drain(&self) -> Vec<TraceEvent> {
        std::mem::take(&mut *self.buffer.lock())
    }

    /// Buffered event count.
    pub fn buffered_count(&self) -> usize {
        self.buffer.lock().len()
    }

    /// Run the processor chain on the buffered events, then send each
    /// processed batch through every exporter. Awaits exporter completion.
    pub async fn flush(self: &Arc<Self>, partial: bool) {
        let events = self.drain();
        if events.is_empty() {
            return;
        }
        let mut stream = events;
        for proc in &self.processors {
            stream = proc.process(stream);
            if stream.is_empty() {
                return;
            }
        }
        let request_id = stream
            .first()
            .map(|e| e.request_id.clone())
            .unwrap_or_default();
        let mut meta = self.metadata.lock().clone();
        meta.partial = partial;
        for exp in &self.exporters {
            let evs = stream.clone();
            exp.export(evs, request_id.clone(), meta.clone()).await;
        }
    }

    /// Fire-and-forget background flush. Used by flush strategies that want
    /// to flush mid-run without blocking the emit path. Returns immediately;
    /// the spawned task drains whatever's buffered when it runs.
    pub fn flush_in_background(self: Arc<Self>, partial: bool) {
        tokio::spawn(async move {
            self.flush(partial).await;
        });
    }
}

/// Convenience builder so engine wiring code can chain processors +
/// exporters without naming `Arc` types.
#[derive(Default)]
pub struct TracePipelineBuilder {
    processors: Vec<Arc<dyn Processor>>,
    exporters: Vec<Arc<dyn Exporter>>,
    flush_strategy: Option<Arc<dyn FlushStrategy>>,
    max_buffered_events: Option<usize>,
}

impl TracePipelineBuilder {
    pub fn processor(mut self, p: Arc<dyn Processor>) -> Self {
        self.processors.push(p);
        self
    }
    pub fn exporter(mut self, e: Arc<dyn Exporter>) -> Self {
        self.exporters.push(e);
        self
    }
    pub fn flush_strategy(mut self, s: Arc<dyn FlushStrategy>) -> Self {
        self.flush_strategy = Some(s);
        self
    }
    pub fn max_buffered_events(mut self, n: usize) -> Self {
        self.max_buffered_events = Some(n);
        self
    }
    pub fn build(self) -> TracePipeline {
        TracePipeline {
            processors: self.processors,
            exporters: self.exporters,
            flush_strategy: self.flush_strategy.unwrap_or_else(|| Arc::new(AtScheduledExit)),
            max_buffered_events: self.max_buffered_events.unwrap_or(100_000),
            buffer: Mutex::new(Vec::new()),
            overflow_warned: Mutex::new(false),
            metadata: Mutex::new(ExportMetadata::default()),
        }
    }
}

/// Suppress an `unused` lint on the `Value` import; ExportMetadata fields
/// are public and might want a `Value`-typed `extras` slot in future.
#[allow(dead_code)]
fn _values_kept_for_extras_extension() -> Value {
    Value::Null
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::tracing::events::{EventKind, TraceEvent};
    use chrono::Utc;
    use std::collections::BTreeMap;

    fn ev(seq: u64) -> TraceEvent {
        TraceEvent {
            event_id: format!("e{seq}"),
            request_id: "r1".into(),
            kind: EventKind::OpStart,
            op_name: Some("op".into()),
            ctx: vec!["main".into()],
            timestamp: Utc::now(),
            seq,
            payload: BTreeMap::new(),
        }
    }

    #[test]
    fn push_appends_to_buffer() {
        let p = TracePipeline::new();
        p.push(ev(0));
        p.push(ev(1));
        assert_eq!(p.buffered_count(), 2);
    }

    #[test]
    fn drain_resets_buffer() {
        let p = TracePipeline::new();
        p.push(ev(0));
        let drained = p.drain();
        assert_eq!(drained.len(), 1);
        assert_eq!(p.buffered_count(), 0);
    }

    struct PassThroughProc;
    impl Processor for PassThroughProc {
        fn process(&self, events: Vec<TraceEvent>) -> Vec<TraceEvent> {
            events
        }
    }

    struct CountingExporter {
        count: Arc<Mutex<u32>>,
    }
    #[async_trait]
    impl Exporter for CountingExporter {
        async fn export(
            &self,
            events: Vec<TraceEvent>,
            _request_id: String,
            _metadata: ExportMetadata,
        ) {
            *self.count.lock() += events.len() as u32;
        }
    }

    #[tokio::test]
    async fn flush_runs_processors_then_exporters() {
        let count = Arc::new(Mutex::new(0u32));
        let pipeline = Arc::new(
            TracePipeline::builder()
                .processor(Arc::new(PassThroughProc))
                .exporter(Arc::new(CountingExporter {
                    count: count.clone(),
                }))
                .build(),
        );
        pipeline.push(ev(0));
        pipeline.push(ev(1));
        pipeline.flush(false).await;
        assert_eq!(*count.lock(), 2);
        assert_eq!(pipeline.buffered_count(), 0);
    }
}
